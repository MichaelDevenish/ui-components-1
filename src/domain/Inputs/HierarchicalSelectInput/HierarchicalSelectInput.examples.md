HierarchicalSelectInput is a special variant of SelectInput for handling hierarchical sets of data.
It requires that the set of data be fully loaded first on the frontend (no async support).

```jsx
initialState = { value: '' };

<HierarchicalSelectInput
  placeholder='Select an option!'
  name='testInput'
  value={state.value}
  hierarchicalOptions={[
    {
      label: 'Australia',
      value: 10
    },
    {
      label: 'Queensland',
      value: 11,
      parentValue: 10
    },
    {
      label: 'NSW',
      value: 12,
      parentValue: 10
    },
    {
      label: 'United States',
      value: 20
    },
    {
      label: 'Victoria',
      value: 13,
      parentValue: 10
    },
    {
      label: 'New York',
      value: 21,
      parentValue: 20
    },
    {
      label: 'Texas',
      value: 22,
      parentValue: 20
    },
    {
      label: 'California',
      value: 23,
      parentValue: 20
    },
    {
      label: 'Brisbane',
      value: 113,
      parentValue: 11
    }
  ]}
  onChange={value => setState({ value })}
/>
```

They have a disabled state:

```jsx
<HierarchicalSelectInput
  name='testInput'
  value={11}
  isDisabled
  hierarchicalOptions={[
    {
      label: 'Parent Option 1',
      value: 10
    },
    {
      label: 'Child Option 1',
      value: 11,
      parentValue: 10
    },
    {
      label: 'Child Option 2',
      value: 12,
      parentValue: 10
    }
  ]}
/>
```

And an invalid state:

```jsx
<HierarchicalSelectInput
  name='testInput'
  value={11}
  isInvalid
  hierarchicalOptions={[
    {
      label: 'Parent Option 1',
      value: 10
    },
    {
      label: 'Child Option 1',
      value: 11,
      parentValue: 10
    },
    {
      label: 'Child Option 2',
      value: 12,
      parentValue: 10
    }
  ]}
/>
```

#### Multiselect Input

They can also be used as a multiselect. In this case, their value is a `string[]` array:

```jsx

initialState = { value: [] };

<HierarchicalSelectInput
  name='testInput'
  value={state.value}
  hierarchicalOptions={[
    {
      label: 'Parent Option 1',
      value: 10
    },
    {
      label: 'Child Option 1',
      value: 11,
      parentValue: 10
    },
    {
      label: 'Child Option 2',
      value: 12,
      parentValue: 10
    }
  ]}
  onChange={value => setState({ value })}
  isMultiSelect
/>
```
