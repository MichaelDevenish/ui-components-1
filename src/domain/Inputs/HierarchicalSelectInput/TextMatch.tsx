import React from 'react'
import { ITextProps } from 'src/domain/Typographies/Text/Text'

import { Text } from '../../Typographies/Text'

const textMatch = (mainText: string, searchText?: string) => {
  if (searchText) {
    const matches = []
    let usedStringLength = 0
    const matcher = mainText.matchAll(new RegExp(
      searchText
        .replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
        .replace(/ /g, '|'),
      'ig'
    ))

    for (const match of matcher) {
      const left = mainText.substring(usedStringLength, (match.index ?? 0))
      usedStringLength = (match.index ?? 0) + match[0].length

      if (left !== '') {
        matches.push({ text: left, isMatch: false })
      }
      matches.push({ text: match[0], isMatch: true })
    }
    if (usedStringLength !== mainText.length) {
      matches.push({ text: mainText.substring(usedStringLength), isMatch: false })
    }

    return matches
  }

  return [{ text: mainText, isMatch: false }]
}

const TextMatch: React.FC<{
  mainText?: string | null
  searchText?: string
  matchTextProps: ITextProps
  otherTextProps: ITextProps
}> = ({ mainText, searchText, matchTextProps, otherTextProps }) => {
  const matchedText = mainText ? textMatch(mainText, searchText) : []
  return (
    <>
      {matchedText.map((match, index) => {
        if (match.isMatch) {
          return (
            <Text isInline {...matchTextProps} key={index}>{match.text}</Text>
          )
        }

        return (
          <Text isInline {...otherTextProps} key={index}>{match.text}</Text>
        )
      })}
    </>
  )
}

export {
  TextMatch
}
