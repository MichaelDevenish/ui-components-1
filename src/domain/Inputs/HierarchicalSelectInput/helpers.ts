import {
  Option,
  OptionValues
} from 'react-select'

interface IHierarchicalOption extends Option {
  parentValue?: OptionValues | null
}

interface IFlattenedOption extends IHierarchicalOption {
  parentOption: IFlattenedOption | null
  depth: number
}

const MAX_SELECT_DEPTH = 8

const flattenHierarchicalOptionsToArray = (
  parentOption: IFlattenedOption | null,
  optionsList: readonly IHierarchicalOption[],
  optionValueChildrenMap: Map<OptionValues, IHierarchicalOption[]>,
  depth: number
) => {
  if (depth > MAX_SELECT_DEPTH) {
    return []
  }

  const flattenedOptions: IFlattenedOption[] = []

  for (const option of optionsList) {
    const flattenedOption = {
      ...option,
      parentOption,
      depth
    }

    flattenedOptions.push(flattenedOption)
    const children = optionValueChildrenMap.get(option.value!)
    if (children) {
      flattenedOptions.push(...flattenHierarchicalOptionsToArray(flattenedOption, children, optionValueChildrenMap, depth + 1))
    }
  }

  return flattenedOptions
}

const convertHierarchicalOptionsToFlattenedOptions = (hierarchicalOptions: readonly IHierarchicalOption[]) => {
  const topLevelOptions = []
  const optionValueChildrenMap = new Map<OptionValues, IHierarchicalOption[]>()

  for (const option of hierarchicalOptions) {
    if (option.value === undefined || option.value === null) {
      throw new Error('Options with undefined or null value are not supported in HierarchicalSelectInput')
    }

    if (option.parentValue === undefined || option.parentValue === null) {
      topLevelOptions.push(option)
    } else {
      const children = optionValueChildrenMap.get(option.parentValue) || []
      optionValueChildrenMap.set(option.parentValue, [...children, option])
    }
  }

  return flattenHierarchicalOptionsToArray(null, topLevelOptions, optionValueChildrenMap, 0)
}

export {
  convertHierarchicalOptionsToFlattenedOptions,
  IHierarchicalOption,
  IFlattenedOption
}
