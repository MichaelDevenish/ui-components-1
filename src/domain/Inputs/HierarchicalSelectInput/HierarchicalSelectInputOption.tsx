import React from 'react'
import Select, {
  OptionComponentProps
} from 'react-select'
import styled from 'styled-components'

import { Props, Variables } from '../../../common'
import { styleForTypographyType } from '../../Typographies/services/textStyles'
import { TextMatch } from './TextMatch'

const StyledOptionLabelWrapper = styled.div`
  margin-right: ${Variables.Spacing.sMedium}px;
  text-overflow: ellipsis;
`

const StyledOptionParentLabelWrapper = styled.div`
  word-break: break-all;
  ${styleForTypographyType(Props.TypographyType.Small)}
`

interface IOption {
  label: string
  parentOption?: IOption
}

const parentDisplay = (option: IOption): string => {
  if (!option.parentOption) {
    return option.label
  }

  return `${parentDisplay(option.parentOption)}/${option.label}`
}

const HierarchicalSelectInputOption: React.FC<OptionComponentProps> = (props) => {
  const {
    option,
    inputValue
  } = props

  const OptionComponent = (Select as any).Option

  return (
    <OptionComponent
      {...props}
      option={option}
    >
      {option.parentOption && (
        <StyledOptionParentLabelWrapper>
          {parentDisplay(option.parentOption)}
        </StyledOptionParentLabelWrapper>
      )}
      <StyledOptionLabelWrapper>
        <TextMatch
          otherTextProps={{ type: Props.TypographyType.Small, color: Variables.Color.n800 }}
          matchTextProps={{ type: Props.TypographyType.Small, weight: Variables.FontWeight.fwSemiBold }}
          mainText={option.label}
          searchText={inputValue}
        />
      </StyledOptionLabelWrapper>
    </OptionComponent>
  )
}

export {
  HierarchicalSelectInputOption
}
