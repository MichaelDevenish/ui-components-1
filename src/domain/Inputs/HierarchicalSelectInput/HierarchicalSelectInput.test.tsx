import { shallow } from 'enzyme'
import React from 'react'

import { HierarchicalSelectInput } from './HierarchicalSelectInput'

describe('<HierarchicalSelectInput />', () => {
  it('should render a select box with options', () => {
    const wrapper = shallow(
      <HierarchicalSelectInput
        name='test'
        value=''
        hierarchicalOptions={[
          {
            label: 'Parent Option 1',
            value: 10
          },
          {
            label: 'Child Option 1',
            value: 11,
            parentValue: 10
          },
          {
            label: 'Child Option 2',
            value: 12,
            parentValue: 10
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a multi select box with options', () => {
    const wrapper = shallow(
      <HierarchicalSelectInput
        name='test'
        value={[]}
        hierarchicalOptions={[
          {
            label: 'Parent Option 1',
            value: 10
          },
          {
            label: 'Child Option 1',
            value: 11,
            parentValue: 10
          },
          {
            label: 'Child Option 2',
            value: 12,
            parentValue: 10
          }
        ]}
        isMultiSelect
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
