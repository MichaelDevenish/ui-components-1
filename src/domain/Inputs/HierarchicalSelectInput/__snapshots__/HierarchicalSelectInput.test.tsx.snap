// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HierarchicalSelectInput /> should render a multi select box with options 1`] = `
<Select
  arrowRenderer={[Function]}
  autoFocus={false}
  autosize={true}
  backspaceRemoves={true}
  backspaceToRemoveMessage="Press backspace to remove {label}"
  className="react-select-test selectInput"
  clearAllText="Clear all"
  clearRenderer={[Function]}
  clearValueText="Clear value"
  clearable={true}
  closeOnSelect={true}
  deleteRemoves={true}
  delimiter=","
  disabled={false}
  escapeClearsValue={true}
  filterOption={[Function]}
  filterOptions={[Function]}
  id="test"
  ignoreAccents={true}
  ignoreCase={true}
  inputProps={Object {}}
  isLoading={false}
  joinValues={false}
  labelKey="label"
  matchPos="any"
  matchProp="any"
  menuBuffer={0}
  menuRenderer={[Function]}
  multi={true}
  name="test[]"
  noResultsText="No results found"
  onBlurResetsInput={true}
  onChange={[Function]}
  onCloseResetsInput={true}
  onSelectResetsInput={true}
  openOnClick={true}
  openOnFocus={false}
  optionComponent={[Function]}
  options={
    Array [
      Object {
        "depth": 0,
        "label": "Parent Option 1",
        "parentOption": null,
        "value": 10,
      },
      Object {
        "depth": 1,
        "label": "Child Option 1",
        "parentOption": Object {
          "depth": 0,
          "label": "Parent Option 1",
          "parentOption": null,
          "value": 10,
        },
        "parentValue": 10,
        "value": 11,
      },
      Object {
        "depth": 1,
        "label": "Child Option 2",
        "parentOption": Object {
          "depth": 0,
          "label": "Parent Option 1",
          "parentOption": null,
          "value": 10,
        },
        "parentValue": 10,
        "value": 12,
      },
    ]
  }
  pageSize={5}
  placeholder="Please Select"
  removeSelected={true}
  required={false}
  rtl={false}
  scrollMenuIntoView={true}
  searchable={true}
  simpleValue={false}
  tabSelectsValue={true}
  trimFilter={true}
  value={Array []}
  valueComponent={[Function]}
  valueKey="value"
/>
`;

exports[`<HierarchicalSelectInput /> should render a select box with options 1`] = `
<Select
  arrowRenderer={[Function]}
  autoFocus={false}
  autosize={true}
  backspaceRemoves={true}
  backspaceToRemoveMessage="Press backspace to remove {label}"
  className="react-select-test selectInput"
  clearAllText="Clear all"
  clearRenderer={[Function]}
  clearValueText="Clear value"
  clearable={true}
  closeOnSelect={true}
  deleteRemoves={true}
  delimiter=","
  disabled={false}
  escapeClearsValue={true}
  filterOption={[Function]}
  filterOptions={[Function]}
  id="test"
  ignoreAccents={true}
  ignoreCase={true}
  inputProps={Object {}}
  isLoading={false}
  joinValues={false}
  labelKey="label"
  matchPos="any"
  matchProp="any"
  menuBuffer={0}
  menuRenderer={[Function]}
  multi={false}
  name="test"
  noResultsText="No results found"
  onBlurResetsInput={true}
  onChange={[Function]}
  onCloseResetsInput={true}
  onSelectResetsInput={true}
  openOnClick={true}
  openOnFocus={false}
  optionComponent={[Function]}
  options={
    Array [
      Object {
        "depth": 0,
        "label": "Parent Option 1",
        "parentOption": null,
        "value": 10,
      },
      Object {
        "depth": 1,
        "label": "Child Option 1",
        "parentOption": Object {
          "depth": 0,
          "label": "Parent Option 1",
          "parentOption": null,
          "value": 10,
        },
        "parentValue": 10,
        "value": 11,
      },
      Object {
        "depth": 1,
        "label": "Child Option 2",
        "parentOption": Object {
          "depth": 0,
          "label": "Parent Option 1",
          "parentOption": null,
          "value": 10,
        },
        "parentValue": 10,
        "value": 12,
      },
    ]
  }
  pageSize={5}
  placeholder="Please Select"
  removeSelected={true}
  required={false}
  rtl={false}
  scrollMenuIntoView={true}
  searchable={true}
  simpleValue={false}
  tabSelectsValue={true}
  trimFilter={true}
  value=""
  valueComponent={[Function]}
  valueKey="value"
/>
`;
